apps/google-analytics-4/src/components/FormControlServiceAccountKey.tsx

```diff
@@ -0,0 +1,275 @@
+import React, { useCallback, useEffect, useState } from 'react';
+import {
+  Collapse,
+  Flex,
+  FormControl,
+  Note,
+  Paragraph,
+  Text,
+  Textarea,
+  TextLink,
+} from '@contentful/f36-components';
+import {
+  CheckCircleIcon,
+  ChevronDownIcon,
+  ChevronRightIcon,
+  ExternalLinkTrimmedIcon,
+} from '@contentful/f36-icons';
+import { css } from 'emotion';
+import { pick, reject } from 'lodash';
+
+import tokens from '@contentful/f36-tokens';
+import InstalledServiceAccountKey from './InstalledServiceAccountKey';
+import type { ServiceAccountKey, ServiceAccountKeyId } from '../types';
+
+interface FormControlState {
+  isInvalid: boolean;
+  errorMessage: string;
+}
+
+interface FormControlServideAccountKeyFileProps {
+  setServiceAccountKey: (serviceAccountKey: ServiceAccountKey | null) => void;
+  setServiceAccountKeyId: (serviceAccountKey: ServiceAccountKeyId | null) => void;
+  setIsInvalid: (isInvalid: boolean) => void;
+  currentServiceAccountKeyId: ServiceAccountKeyId | null;
+  className?: string;
+}
+
+class AssertionError extends Error {}
+
+const styles = {
+  credentialsInput: css({
+    fontFamily: 'monospace',
+  }),
+  credentialsNote: css({
+    marginBottom: tokens.spacingM,
+  }),
+  credentialsExplanation: css({
+    marginTop: tokens.spacingL,
+  }),
+};
+
+const placeholderText = `{
+  "type": "service_account",
+  ...
+}`;
+
+const keysOfServiceAccountKey: Array<keyof ServiceAccountKey> = [
+  'type',
+  'project_id',
+  'private_key_id',
+  'private_key',
+  'client_email',
+  'client_id',
+  'auth_uri',
+  'token_uri',
+  'auth_provider_x509_cert_url',
+  'client_x509_cert_url',
+];
+
+const convertKeyFileToServiceAccountKey = (keyFile: string): ServiceAccountKey => {
+  const parsedKeyFile = JSON.parse(keyFile);
+  assertServiceAccountKey(parsedKeyFile);
+
+  // ensure key file never contains extraneous keys
+  return pick(parsedKeyFile, keysOfServiceAccountKey);
+};
+
+const convertServiceAccountKeyToServiceAccountKeyId = (
+  serviceAccountKey: ServiceAccountKey
+): ServiceAccountKeyId => ({
+  id: serviceAccountKey.private_key_id,
+  clientEmail: serviceAccountKey.client_email,
+  clientId: serviceAccountKey.client_id,
+  projectId: serviceAccountKey.project_id,
+});
+
+function assertServiceAccountKey(value: any): asserts value is ServiceAccountKey {
+  if (value?.type !== 'service_account')
```

Comment:
nit: I think we normally use parenthesis for single-line blocks in this repo. But maybe it's a mix and I just don't know.

Reply:
👍 Makes sense. I'm trying to remember if i even did this or it was Prettier 🤔? Either way I personally prefer explicit curly braces so happy to change it if prettier let's me.